// @dynamic-labs/client/src/modules/walletsModule/waasModule/waasModule.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var messageTransport = require('@dynamic-labs/message-transport');

const createWaasModule = (core) => {
    const requestChannel = messageTransport.createRequestChannel(core.messageTransport);
    return {
        backupKeySharesToGoogleDrive: (walletId, args) => requestChannel.request('waas_backupKeySharesToGoogleDrive', walletId, args),
        delegateKeyShares: (walletId, args) => requestChannel.request('waas_delegateKeyShares', walletId, args),
        exportClientKeyshares: (walletId, args) => requestChannel.request('waas_exportClientKeyshares', walletId, args),
        importPrivateKey: (walletId, args) => requestChannel.request('waas_importPrivateKey', walletId, args),
        refreshWalletAccountShares: (walletId, args) => requestChannel.request('waas_refreshWalletAccountShares', walletId, args),
        signRawMessage: (walletId, args) => requestChannel.request('waas_signRawMessage', walletId, args),
        updatePassword: (walletId, args) => requestChannel.request('waas_updatePassword', walletId, args),
    };
};

exports.createWaasModule = createWaasModule;
