// @dynamic-labs/client/src/utils/setupFetchHandler/setupFetchHandler.js
'use client'
import { __awaiter } from '../../../_virtual/_tslib.js';
import { createRequestChannel } from '@dynamic-labs/message-transport';
import { logger } from '../logger/logger.js';

const setupFetchHandler = (core) => {
    const fetchRequestChannel = createRequestChannel(core.messageTransport);
    fetchRequestChannel.handle('fetch', (input, init) => __awaiter(void 0, void 0, void 0, function* () {
        try {
            const response = yield fetch(input, init);
            const responseBody = yield response.text();
            return {
                body: responseBody,
                // eslint-disable-next-line @typescript-eslint/no-explicit-any
                headers: Object.fromEntries(response.headers.entries()),
                status: response.status,
                statusText: response.statusText,
            };
        }
        catch (error) {
            logger.error('Fetch request failed in setupFetchHandler', {
                error: error instanceof Error ? error.message : String(error),
                input: typeof input === 'string' ? input : input.url,
                method: (init === null || init === void 0 ? void 0 : init.method) || 'GET',
                stack: error instanceof Error ? error.stack : undefined,
            });
            throw error;
        }
    }));
};

export { setupFetchHandler };
