// @dynamic-labs/client/src/modules/mfaModule/mfaModule.js
'use client'
import { createRequestChannel } from '@dynamic-labs/message-transport';

const createMfaModule = (core) => {
    const requestChannel = createRequestChannel(core.messageTransport);
    return {
        addDevice: (...args) => requestChannel.request('addDevice', ...args),
        authDevice: (...args) => requestChannel.request('authDevice', ...args),
        authRecoveryCode: (...args) => requestChannel.request('authRecoveryCode', ...args),
        authenticateDevice: (...args) => requestChannel.request('authenticateDevice', ...args),
        completeAcknowledgement: (...args) => requestChannel.request('completeAcknowledgement', ...args),
        deleteUserDevice: (...args) => requestChannel.request('deleteUserDevice', ...args),
        getRecoveryCodes: (...args) => requestChannel.request('getRecoveryCodes', ...args),
        getUserDevices: (...args) => requestChannel.request('getUserDevices', ...args),
        updateUserDevice: (...args) => requestChannel.request('updateUserDevice', ...args),
        verifyDevice: (...args) => requestChannel.request('verifyDevice', ...args),
    };
};

export { createMfaModule };
