// @dynamic-labs/client/src/modules/mfaModule/mfaModule.cjs
'use client'
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var messageTransport = require('@dynamic-labs/message-transport');

const createMfaModule = (core) => {
    const requestChannel = messageTransport.createRequestChannel(core.messageTransport);
    return {
        addDevice: (...args) => requestChannel.request('addDevice', ...args),
        authDevice: (...args) => requestChannel.request('authDevice', ...args),
        authRecoveryCode: (...args) => requestChannel.request('authRecoveryCode', ...args),
        authenticateDevice: (...args) => requestChannel.request('authenticateDevice', ...args),
        completeAcknowledgement: (...args) => requestChannel.request('completeAcknowledgement', ...args),
        deleteUserDevice: (...args) => requestChannel.request('deleteUserDevice', ...args),
        getRecoveryCodes: (...args) => requestChannel.request('getRecoveryCodes', ...args),
        getUserDevices: (...args) => requestChannel.request('getUserDevices', ...args),
        updateUserDevice: (...args) => requestChannel.request('updateUserDevice', ...args),
        verifyDevice: (...args) => requestChannel.request('verifyDevice', ...args),
    };
};

exports.createMfaModule = createMfaModule;
